File: tw/src/main/java/pl/edu/agh/philosophers/WaiterFork.java
package pl.edu.agh.philosophers;

import pl.edu.agh.util.Utils;

import java.util.List;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class WaiterFork implements Fork {
    private final Lock lock = new ReentrantLock();
    private List<Boolean> occupied;
    private List<Condition> forks;
    private Condition waiter;
    private int howManyPresent;

    public WaiterFork() {
        occupied = Utils.createObjects(5, () -> false);
        forks = Utils.createObjects(5, lock::newCondition);
        waiter = lock.newCondition();
        howManyPresent = 0;
    }

    @Override
    public void take(int i) {
        try {
            lock.lock();
            while (howManyPresent == 4) {
                waiter.await();
            }
            howManyPresent++;
            while (occupied.get(i)) {
                forks.get(i).await();
            }
            occupied.set(i, true);
            while (occupied.get((i + 1) % 5)) {
                forks.get((i + 1) % 5).await();
            }
            occupied.set((i + 1) % 5, true);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        } finally {
            lock.unlock();
        }
    }

    @Override
    public void putBack(int i) {
        try {
            lock.lock();
            occupied.set(i, false);
            forks.get(i).signal();
            occupied.set((i + 1) % 5, false);
            forks.get((i + 1) % 5).signal();
            howManyPresent--;
            waiter.signal();
        } finally {
            lock.unlock();
        }
    }
}

File: philosopers_pyt/waiter_fork.py
import threading


class WaiterFork:
    def __init__(self):
        self.lock = threading.RLock()
        self.occupied = list(map(lambda _: False, range(5)))
        self.forks = list(map(lambda _: threading.Condition(lock=self.lock), range(5)))
        self.waiter = threading.Condition(lock=self.lock)
        self.how_many_present = 0

    def take(self, i):
        try:
            self.lock.acquire()
            while self.how_many_present == 4:
                self.waiter.wait()
            self.how_many_present += 1
            while self.occupied[i]:
                self.forks[i].wait()
            self.occupied[i] = True
            while self.occupied[(i + 1) % 5]:
                self.forks[(i + 1) % 5].wait()
            self.occupied[(i + 1) % 5] = True
        finally:
            self.lock.release()

    def put_back(self, i):
        try:
            self.lock.acquire()
            self.occupied[i] = False
            self.forks[i].notify()
            self.occupied[(i + 1) % 5] = False
            self.forks[(i + 1) % 5].notify()
            self.how_many_present -= 1
            self.waiter.notify()
        finally:
            self.lock.release()
